### PR TITLE
Library refactor + Multiple formats

### LINKED ISSUE #11: Create json printer
Create an optional json printer instead of using the tree printing

### PR DESCRIPTION
**Pull request by chentex**
Changes:
* Create library and move execution code to the library
* Adds formatter to generate different outputs
* Creates a Result type
* Allows to be used as a library

New Flags:
* format [plain, stdout, json, yaml]
* filename [to save results to a file]

Missing Changes:
* Walking the Objects it's failing, so I need to check before adding that to the results.

fixes #11

### CODE REVIEW

**Comment by rikatz** [comment]
So I finally got some vacations and will be able to take a look into this :D

I'll start reviewing this afternoon and if there's nothing critical I'll merge this tomorrow

**Comment by rikatz** [comment]
@chentex Overall you've made an amazing job!

I've left some comments here about the Kubernetes config files, and also some field naming.

I'll take a look later today into why maybe the walker isn't working as expected :)

If you wan't to discuss something please feel free to also ping me on Kubernetes Slack :) 

Also as a suggestion, I think you could put the diagram posted in https://github.com/rikatz/kubepug/issues/11#issuecomment-592551997 in a later commit, as a documentation improvement

**Comment by chentex** [comment]
> @chentex Overall you've made an amazing job!
> 
> I've left some comments here about the Kubernetes config files, and also some field naming.
> 

Thanks I will take a look of them today.

**Comment by rikatz** [comment]
@chentex as I'm not looking always Github, please give me a ping whenever you're ready to another review.

Thank you!!

**Comment by rikatz** [comment]
@chentex friendly ping :D

File: pkg/kubepug/types.go [hunk 1]
```diff
@@ -20,3 +12,37 @@ type KubeAPI struct {
...
+	Kind        string           `json,yaml:"kind,omitempty"`
+	Version     string           `json,yaml:"version,omitempty"`
+	Name        string           `json,yaml:"name,omitempty"`
+	Deprecated  bool             `json,yaml:"deprecated,omitempty"`
+	Items       []DeprecatedItem `json,yaml:"deprecated_items,omitempty"`
+}
+
+// DeprecatedItem definition of the Items inside a deprecated API
+type DeprecatedItem struct {
+	Kind      string `json,yaml:"kind,omitempty"`
```
[the remark this answers is not in the document]
Comment (chentex) [reply pkg/kubepug/types.go]: Changed the name for this.

File: pkg/kubepug/types.go [hunk 2]
```diff
@@ -20,3 +12,37 @@ type KubeAPI struct {
...
+	Name      string `json,yaml:"name,omitempty"`
```
[the remark this answers is not in the document]
Comment (chentex) [reply pkg/kubepug/types.go]: the same here

File: cmd/kubepug.go [hunk 3]
```diff
@@ -72,8 +73,6 @@ func init() {
 		rootCmd.Example = cmdValue
 	}
 
-	kubernetesConfigFlags = genericclioptions.NewConfigFlags(true)
-	kubernetesConfigFlags.AddFlags(rootCmd.Flags())
```
[the remark this answers is not in the document]
Comment (chentex) [reply cmd/kubepug.go]: I included config Flags as a parameter, if feel it is the best way to go and keep it configurable.

**Comment by chentex** [comment]
I rebased averything from master, it is ready for a new round of review

**Comment by rikatz** [comment]
lgtm

I've found some performance difference between the old and new behavior, and also walker marking everything as deprecated :)

But as the output formatter is working I'm going to merge this PR and not cut a new release until we can solve the performance and walker issues :)

Thank you very much for your hard work @chentex I really appreciate that :)

**Pull request merged**